File: General-DSA-Questions/Stack/6_nextGreaterElementInCircularArray.cpp
// Given a circular array arr[], for each element arr[i], find the next greater element, wrapping around to the beginning if needed.
// If no such element exists, return -1.

// Approach:
// 1. We can use similar logic to the next greater element problem, but we need to handle the circular nature of the array.
// 2. We can simulate the circular array by iterating through the array twice.
// 3. We use modulo (i % n) to wrap around.

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

vector<int> nextGreaterElementsCircular(const vector<int>& nums) {
    int n = nums.size();
    vector<int> result(n, -1);
    stack<int> s; // will store indices

    for (int i = 2 * n - 1; i >= 0; --i) {  // Iterate twice to simulate circular array
        // Calculate the index in the original array
        // i % n gives us the index in the original array
        // This way, we can handle the circular nature of the array
        int idx = i % n;

        // Pop elements <= current one
        while (!s.empty() && nums[s.top()] <= nums[idx]) {
            s.pop();
        }

        if (i < n) { // Only fill result in the first pass
            if (!s.empty()) {
                result[idx] = nums[s.top()];
            }
        }

        s.push(idx); // Always push current index
    }

    return result;
}

int main() {
    vector<int> nums = {1, 2, 1};
    vector<int> result = nextGreaterElementsCircular(nums);

    cout << "Next Greater in Circular Array:\n";
    for (int val : result) {
        cout << val << " ";
    }

    return 0;
}

File: JS-DSA-Questions/Stack/6_nextGreaterElementInCircularArray.js
// Given a CIRCULAR array arr[], for each element find the next greater element, wrapping around
// to the beginning if needed. If none exists, use -1.

// Approach:
// 1. Same monotonic-stack logic as the linear next-greater-element problem.
// 2. Simulate the circular array by iterating twice (indices 2n-1 down to 0).
// 3. Use i % n to map back to the original index.

// JavaScript note: the stack stores indices; a plain array serves as the stack.
function nextGreaterElementsCircular(nums) {
  const n = nums.length;
  const result = new Array(n).fill(-1);
  const s = []; // stores indices

  for (let i = 2 * n - 1; i >= 0; i--) {
    const idx = i % n;

    // Pop elements <= current
    while (s.length > 0 && nums[s[s.length - 1]] <= nums[idx]) {
      s.pop();
    }

    if (i < n) {
      // Only fill result on the first (real) pass
      if (s.length > 0) {
        result[idx] = nums[s[s.length - 1]];
      }
    }

    s.push(idx);
  }

  return result;
}

console.log("Next Greater in Circular Array:", nextGreaterElementsCircular([1, 2, 1]).join(" ")); // 2 -1 2
